### DISCUSSION TITLE
Using ECR as a Private Image Registry

### DISCUSSION THREAD

**Discussion by metaskills** [community/community#5873] [Codespaces] [answered]
I've read this page (https://docs.github.com/en/codespaces/codespaces-reference/allowing-your-codespace-to-access-a-private-image-registry#common-image-registry-servers) several times and gotten Docker Hub working correctly, but I can not get ECR working. I've created several users and IAM credentials using full access to ECR and still can not get the image to be read in via Docker or a service via `dockerComposeFile`. 

Can someone confirm that values for each? For example, the docs say the server should be in the format of `<aws_account_id>.dkr.ecr.<region>.amazonaws.com` and I have done that. Am I correct that the `<*>_CONTAINER_REGISTRY_USER` is the Access Key ID and `<*>_CONTAINER_REGISTRY_PASSWORD` is the secret key? Also, has anyone gotten this to work for ECR?

**Comment by asciimike** [comment]
I don't believe that AWS ECR supports using the Access Key ID/secret key paradigm, I think you have to generate a short lived authorization token and refresh it (per [their docs](https://docs.aws.amazon.com/AmazonECR/latest/userguide/registry_auth.html)). We are planning on offering a secrets API (like [what Actions provides](https://docs.github.com/en/rest/reference/actions#secrets)) to enable you to programmatically refresh that token.

**Comment by metaskills** [comment]
So @asciimike are you saying that the feature on [this page](https://docs.github.com/en/codespaces/codespaces-reference/allowing-your-codespace-to-access-a-private-image-registry#common-image-registry-servers) with regard to ECR a) does not work or b) will if I or Codespaces does something different? 

My expectation was that given the key and secret I provided, Codespaces would have done the API calls needed to do the ECR login. For example, we use this GitHub Action (https://github.com/marketplace/actions/docker-login#aws-elastic-container-registry-ecr) and we already did the work to make sure the key/secret has IAM permissions to do `sts:GetServiceBearerToken` and everything in the Action just works. 

In fact, I think we recently had to fix a bug where we had to add that `sts:GetServiceBearerToken` capabilities to our user. Let me test again real quick.

**Reply by asciimike** [reply]
I don't believe we do the credential exchange automatically, I assume we'll simply pass everything through. Would you mind trying with a short lived token to confirm this? If that works, I can update the docs and then can discuss with the team what would be required to do the credential swap.

EDIT: after a brief consultation of the code, I'm pretty confident we're not doing the token swap automatically, so you'll have to provide the pre-swapped token. I'll raise an issue with the team to discuss how we can make this better.

**Reply by metaskills** [reply]
>  Would you mind trying with a short lived token to confirm this?

If I did and it worked... and given the token would last a max of 12 hours... would it be fair to say that is a broken feature vs a documentation change? Or do you expect folks to put some automation in place to generate/update those org secrets?

**Reply by asciimike** [reply]
Thanks for confirming, unfortunately I think the current behavior is "works as intended" though I recognize that there's a gap in how it works and how it _should_ work.

I have several action items to resolve this:
 - **Immediately:** update the documentation to reflect which token is required, how to acquire one, and the limitations discussed above
 - **Near term:** publish an API that lets customers update secrets ([similar to Actions](https://docs.github.com/en/rest/reference/actions#secrets)), which would alleviate the immediate pain of having to manually update the credential in the UI every 12 hours (which, I have no expectation of folks being able to/willing to do). There is reasonable signal from a similar issue in Dependabot that this might be sufficient (https://github.com/dependabot/dependabot-core/issues/3297)
 - **Longer term:** consider if we want to implement the token swap ourselves, if we can provide an Action to do it, etc. 

How does this plan sound?

**Reply by metaskills** [reply]
Hey @asciimike I appreciate the work! I would like to emphasis this is a big pain point now. Feedback on each of your items in order of importance below.

> Near term: publish an API that lets customers update secrets

I need this now. It is hard to tell the organization we can use Codespaces while someone has to go into and do some clicking to reset a token every 12 hours. It is hard for me to describe that as working. Is there an ETA on the API interface? For context, we have found our CI/CD images really useful for adopting Codespaces in legacy applications and those are pulled from ECR when building the development container.

> Immediately: update the documentation 

I did find this a tad bit useful but it took me a long time to get it right. Here are some things I learned along the way that may help with the documentation.

1. The matching `<*>_CONTAINER_REGISTRY_USER` has to be `AWS`.
2. I do not HAVE to use `ECR_CONTAINER_REGISTRY_PASSWORD` I can use the `<*>_CONTAINER_REGISTRY_PASSWORD` pattern.
3. The `<*>_CONTAINER_REGISTRY_PASSWORD` value is the output of `aws ecr get-login-password --region us-east-1`.
4. The `<*>_CONTAINER_REGISTRY_SERVER` is as the docs described. Ex: `12345678901.dkr.ecr.us-east-1.amazonaws.com`

> Longer term: consider if we want to implement the token swap ourselves

Honestly, not a lot of people are going to use this feature if they have to setup some outside function to change the `<*>_CONTAINER_REGISTRY_PASSWORD` every 12 hours. I'm willing to do it, but others may not. Hope that helps y'all make decisions on where to do work... and also... THANK YOU FOR CODESPACES! 🖤

**Reply by asciimike** [reply]
On a plane so the first response will be a bit terse, will follow up.

We've released an API to set secrets for a user: https://docs.github.com/en/rest/reference/codespaces#create-or-update-a-secret-for-the-authenticated-user

Mind giving that a try and seeing if it works for you? Per repo and org secrets are on their way, as I recognize that you probably want to set this for all users using a repo.

Really appreciate the docs updates, I'm going to be making some updates to that section soon, and I agree that we probably need to do the credential swap for folks; I've started talking to engineering about how much effort this is to build/maintain.

**Reply by metaskills** [reply]
> Per repo and org secrets are on their way, as I recognize that you probably want to set this for all users using a repo.

Yes... and since @Chuxel neat temp solution is not working... it would be amazing if those were in the near future?

**Reply by asciimike** [reply]
Ok, thinking a bit longer term about how we might make this better, I think that the best solution we can provide is:
 - you set the following secrets: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`
 - you set a `*_CONTAINER_REGISTRY_SERVER` with your ECR domain
 - we'll do a credential exchange via this [API](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html) (which returns a base64 encoded credential) when the codespace is started
 - we'll populate `*_CONTAINER_REGISTRY_USER=AWS` and `*_CONTAINER_REGISTRY_PASSWORD=base64_decode(credential)`
 - we'll run `echo $*_CONTAINER_REGISTRY_PASSWORD | docker login --username $*_CONTAINER_REGISTRY_USER --password-stdin $*_CONTAINER_REGISTRY_SERVER` to log you in

I think another advantage of this solution is that it makes other operations with the AWS CLI easier, since you'll have those credentials readily available (potentially in the future we might even want to write them to `~/.aws/credentials`).

@metaskills how does this sound to you?

**Reply by metaskills** [reply]
I think that sounds "fine" however I did want to call out two important points.

1. As is now, other `*_CONTAINER_REGISTRY_*` ENV variables that match the private registry pattern are not exposed to the user. I kind of like that. Means I can add secrets that have a functional output without giving them to users. Would your change be the exception to that rule now? Or am I remember the existing behavior incorrectly?
2. When I tried to DIY my own `docker login ...` method it broke our Codespaces. I mentioned this here. https://github.com/github/feedback/discussions/5873#discussioncomment-1564644 and was never able to verify why. I did not push further on it because I really liked how point (1) above worked and did not want to expose Codespaces ECR integration secrets with every user.

**Reply by asciimike** [reply]
We are not planning to expose the `*_CONTAINER_REGISTRY_*` variables, so the behavior will remain the same. The `AWS_*` ones would be regular secrets, and thus accessible by folks though (presumably people will want to use them for other tools).

**Reply by metaskills** [reply]
> you set the following secrets: AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY
> The AWS_* ones would be regular secrets...

I missed this the first time around... when you say `AWS_ACCESS_KEY_ID` & `AWS_SECRET_ACCESS_KEY` I assume you mean at the organization level? If so, that sounds like a bad idea. It would set this very specific use case (ECR Login) to a very general (and global) AWS credential namespace. That feels like a leaky abstraction. As an organization, what I would like is to be able to do is set `*_CONTAINER_REGISTRY_*` ENVs for ECR and have everything done for me and hidden from users. I feel open to solutions but not ones that leak into the user accessible ENV vars. Especially if those ENV vars could conflict with other use cases.

**Reply by asciimike** [reply]
You could provide it at a user or repo level to apply for a subset of users.

My initial thought was that ECR is a subset of the AWS CLI, and that having the AWS CLI credentials be plumbed through would be generally useful, as I've had requests for providing `az` and `gcloud` credentials as well. I recognize that these may make more sense to be set individually though, as it's a slightly different use case.

As I understand it, your preference would be to have `*_CONTAINER_REGISTRY_USER=AWS_ACCESS_KEY_ID` and `*_CONTAINER_REGISTRY_PASSWORD= AWS_SECRET_ACCESS_KEY`?

**Reply by metaskills** [reply]
> As I understand it, your preference would be to have *_CONTAINER_REGISTRY_USER=AWS_ACCESS_KEY_ID and *_CONTAINER_REGISTRY_PASSWORD=AWS_SECRET_ACCESS_KEY?

That is correct. That would be the expected behavior and the least DX friction. The documentation could mention that this user needs the `sts:GetServiceBearerToken` to work. As is now I am doing a hacky rotate of the key/secret every hour via the worlds worst automation and duct tape. I would imagine I am the only person doing this for ECR integration and that changing the interface around to not require posting new keys every 12 hours would be a welcome change for anyone wanting the as advertised private registry integration.

**Reply by asciimike** [reply]
I agree, my only concern was folks would think that the user has to be `aws` rather than letting us do the key exchange. Unfortunately, this is probably going to be one of those features where you are forced to read the docs to see how it works.

I do think that we'll offer easier credential access for AWS/GCP/Azure tooling in the future though, since I'm seeing a lot of questions come up around this, but it likely be distinct from this feature.

**Comment by Chuxel** [comment]
@metaskills I have a near term workaround that takes advantage of `initalizeCommand` in `devcontainer.json` and uses the AWS CLI to sign in whenever a codespaces is created.

See https://github.com/chuxel/codespaces-ecr

There's slightly different secret names you set, and you can customize the names of the secrets in the event you need more than one across repos, etc.

We need to get first class support in place here, but I think this should unblock you.  /cc: @asciimike

**Reply by metaskills** [reply]
@asciimike Thanks so much for sharing that user scoped Codespaces API, those will come in handy in the future. I've marked this work around as the answer because it solves the problem really well now till either a) an org level API is created or b) automatic bearer token is done for us.

**Reply by metaskills** [reply]
Well actually, I'm finding that this pulls the image but the VS Code server fails after it. I've tried this several times and for some reason this method messes up the container so bad it won't even do a recover mode. 

<img width="483" alt="Screen Shot 2021-10-30 at 11 40 13 PM" src="https://user-images.githubusercontent.com/2381/139566349-167abab7-be8c-4c3e-b6e7-925b638e33d1.png">

**Reply by Chuxel** [reply]
Hmmm. That's odd. Can you share your devcontainer.json?

**Reply by metaskills** [reply]
I simplified it a lot while debugging. All the way down to just an "image" property vs a compose, etc. I will email the create log output to Jonathan Carter and he can maybe share with you? Always ended here.

```
...
 * Starting OpenBSD Secure Shell server sshd
docker process exited with exit code 0
(*) Starting ssh server.
```

**Reply by Chuxel** [reply]
Yep that works.  You can also try changing `set -e` to `set +e` in the event the script is erroring out in an odd way.  Not quite sure why this would work in my ECR case, so it's actually possible something unrelated is going on as well. Does the image have a custom entrypoint or command?

**Reply by metaskills** [reply]
I think our script runs fine as the entire dev container is built. Not sure about the entrypoint, but if we are simply changing the method of docker login with the same image, how might it come into play?

**Reply by Chuxel** [reply]
Just adding this to the public issue for others since we've corresponded offline. It is unclear why this would be happening given the image build is complete and initalizeCommand fires on the host VM. Something in this particular image (or possibly dotfiles) seems to be tripping things up. We're having trouble reproducing. If anyone else hits this problem, if you send across a codespace ID, we can take a look at the codespace to dig in further.

**Comment by asciimike** [comment, marked as answer]
Per all the discussion here, we've solved this in a more elegant way which allows you to provide an access key ID and secret. The docs have been updated accordingly: https://docs.github.com/en/codespaces/codespaces-reference/allowing-your-codespace-to-access-a-private-image-registry#accessing-aws-elastic-container-registry

**Reply by metaskills** [reply]
THANK YOU! The docs look good. I did notice there were a few lower case `_container_` that would be more consistent as `_CONTAINER_`.

**Reply by samsarkleio** [reply]
Hi @asciimike,

Thanks for this! I just wanted to double check what the intent and workflow are now after this change.

My assumptions are that if:
- we set `*_CONTAINER_REGISTRY_SERVER=<aws_account_id>.dkr.ecr.<region>.amazonaws.com`
- and we set `*_CONTAINER_REGISTRY_USER = <AWS_ACCESS_KEY_ID>`
- and we set `*_CONTAINER_REGISTRY_PASSWORD = <AWS_SECRET_ACCESS_KEY>`

...then during a Codespace initialization, something equivalent to the following is going to happen:
- `aws ecr get-login-password --region <region> | sudo docker login --username AWS --password-stdin <aws_account_id>.dkr.ecr.<region>.amazonaws.com`.

By equivalent, I mean it may not literally use the `aws` cli, but it will hit the API in some fashion (maybe this [API](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_GetAuthorizationToken.html) as you mentioned above) to get temporary login credentials, and pipe them into `docker login`.

And to make this all work, the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY of the user in question needs to have `sts:GetServiceBearerToken` rights.

Is that correct?

**Reply by asciimike** [reply]
@samsarkleio you're exactly right (and yes, we use the API rather than the CLI). You'll also need to make sure the user in question has read access on the image as well as the token swap permissions.

**Reply by samsarkleio** [reply]
@asciimike Awesome thank you!

One more question, is there any startup/initialization path where the `*_CONTAINER_REGISTRY_*` variables could ever be seen by users? Is it ever exposed in logs, or can it be grabbed from the box itself?

**Reply by asciimike** [reply]
I don't believe these variables are exposed, and even if they were, we'll have swapped them out for the short lived token by the time they could be available, so there isn't a risk of exposing the access key ID/secret key.
